examples/demo2/app.rs

```diff
@@ -1,252 +1,223 @@
-use unicode_width::UnicodeWidthStr;
-
-use crate::{
-    big_text::{BigTextBuilder, PixelSize},
-    Root, Term,
-};
-
-#[derive(Debug)]
+use color_eyre::{eyre::Context, Result};
+use crossterm::event::{Event, KeyCode, KeyEvent, KeyEventKind};
+use itertools::Itertools;
+use ratatui::{prelude::*, widgets::*};
+use strum::{Display, EnumIter, FromRepr, IntoEnumIterator};
+
+use crate::{destroy, tabs::*, term, THEME};
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
 pub struct App {
-    term: Term,
-    context: AppContext,
     mode: Mode,
+    tab: Tab,
+    about_tab: AboutTab,
+    recipe_tab: RecipeTab,
+    email_tab: EmailTab,
+    traceroute_tab: TracerouteTab,
+    weather_tab: WeatherTab,
 }
 
 #[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
 enum Mode {
     #[default]
-    Normal,
+    Running,
     Destroy,
     Quit,
 }
 
-#[derive(Debug, Default, Clone, Copy)]
-pub struct AppContext {
-    pub tab_index: usize,
-    pub row_index: usize,
+#[derive(Debug, Clone, Copy, Default, Display, EnumIter, FromRepr, PartialEq, Eq)]
+enum Tab {
+    #[default]
+    About,
+    Recipe,
+    Email,
+    Traceroute,
+    Weather,
+}
+
+pub fn run(terminal: &mut Terminal<impl Backend>) -> Result<()> {
+    App::new().run(terminal)
 }
 
 impl App {
-    fn new() -> Result<Self> {
-        Ok(Self {
-            term: Term::start()?,
-            context: AppContext::default(),
-            mode: Mode::Normal,
-        })
-    }
-
-    pub fn run() -> Result<()> {
-        install_panic_hook();
-        let mut app = Self::new()?;
-        while !app.should_quit() {
-            app.draw()?;
-            app.handle_events()?;
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Run the app until the user quits.
+    pub fn run(&mut self, terminal: &mut Terminal<impl Backend>) -> Result<()> {
+        while self.is_running() {
+            self.draw(terminal)?;
+            self.handle_events()?;
         }
-        Term::stop()?;
         Ok(())
     }
 
-    fn draw(&mut self) -> Result<()> {
-        self.term
+    fn is_running(&self) -> bool {
+        self.mode != Mode::Quit
+    }
+
+    /// Draw a single frame of the app.
+    fn draw(&self, terminal: &mut Terminal<impl Backend>) -> Result<()> {
+        terminal
             .draw(|frame| {
-                frame.render_widget(Root::new(&self.context), frame.size());
+                frame.render_widget(self, frame.size());
                 if self.mode == Mode::Destroy {
-                    destroy(frame);
+                    destroy::destroy(frame);
                 }
             })
-            .context("terminal.draw")?;
+            .wrap_err("terminal.draw")?;
         Ok(())
     }
 
+    /// Handle events from the terminal.
+    ///
+    /// This function is called once per frame, The events are polled from the stdin with timeout of
+    /// 1/50th of a second. This was chosen to try to match the default frame rate of a GIF.
```

Comment:
Is it the default or maximum framerate?

```suggestion
    /// 1/50th of a second. This was chosen to try to match the maximum frame rate of a GIF.
```

Reply:
changed to "...the default frame rate of a GIF in VHS."